[TECH] Utiliser la librairie openid-client (PIX-10990)

igorissen opened the pull request:
## :unicorn: Problème

Une bonne partie du code API concernant l’authentification OIDC pourrait être remplacé par une librairie standard avec plusieurs avantages :

- moins de code générique à maintenir par Pix
- l’assurance du respect des standards
- plus de sécurité (avec la validation des signatures des tokens notamment)

## :robot: Proposition

Utiliser la librairie [_openid-client_](https://www.npmjs.com/package/openid-client) qui fonctionne en récupérant par le réseau l'_openid-configuration_ de l'OIDC Provider.

En s’appuyant sur 2 nouvelles variables d'ENV : 
* `XXX__OPENID_CONFIGURATION_URL` (où `XXX` est à remplacer par le code de l’OIDC Provider)
* `XXX_REDIRECT_URI` (où `XXX` est à remplacer par le code de l’OIDC Provider)

## :rainbow: Remarques

* :warning: Cette PR nécessite de définir de nouvelles variables d'environnement en intégration, recette et production. 
* Cette PR a plus de 1000 changements (additions + suppressions), mais cette limite max des standards de code a été définie avant le début de la création de cette PR.
* De manière à ne pas alourdir cette PR déjà très grosse, celle-ci ne contient pas la gestion des erreurs OIDC émises par _openid-client_. Mais cela ne représente pas un risque de sécurité : cela génère des erreurs `500` au lieu qu'il y ait un traitement spécifique. C'est très temporaire, une PR ajoutant cette gestion des erreurs arrive juste après.
* Dans plusieurs fichiers de test on a mis des `skip` et on a laissé du code en commentaire pour reprendre cette façon de faire plus tard. En effet : 
   * comme expliqué au-dessus, il manque la gestion des erreurs OIDC émises par _openid-client_
   * lorsque c'était possible on a stubé de manière peu élégante et peu pérenne le `OidcAuthenticationService.client`, mais la vraie bonne solution est de rendre l’_oidcExampleNetProvider_ encore plus proche de la réalité en lui faisant exposer son JSON Web Key Set (JWKS) et en lui faisant signer pour de vrai ses ID Tokens, mais ce travail nous prendrait trop de temps pour l'instant et on décide de le faire plus tard

## :100: Pour tester

Bien définir les nouvelles variables d’ENV (voir plus haut la section _Proposition_) : 
* `XXX__OPENID_CONFIGURATION_URL`
* `XXX_REDIRECT_URI`

Tout d'abord vérifier qu'il n'y a pas de régression au niveau des fonctionnalités des SSO. Pour cela réaliser les actions suivantes pour chacun des fournisseurs d'identités : 

1. Se connecter au SSO avec un compte d'utilisateur n'ayant pas déjà accès à Pix
2. Créer son compte
3. Se déconnecter
4. Se reconnecter avec le même compte

Enfin, tester le fait qu'un SSO ne soit pas joignable par le réseau ne stoppe pas le démarrage de Pix API. Pour cela spécifier un mauvais `WELLKNOWN_URL` pour un SSO et constater que Pix API démarre bien et qu'un log d'erreur indiquant le SSO non disponible est émis.

lego-technix committed:
feat(api): improve OIDC Provider has been disabled error message

igorissen committed:
refactor(api): reorder keys

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
refactor(api): reorder constructor parameters

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
refactor(api): reorder imports

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): use "configKey" property

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): add redirectUri & openidConfigurationUrl attributes

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
refactor(api): modify test names

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): add default scope

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
refactor(api): use openid client callback method

to exchange authorization code for token

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

lego-technix committed:
test(api): make createServerWithTestOidcProvider provide an openid configuration

lego-technix committed:
feat(api): add openid-client dependency

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

lego-technix committed:
feat(api): add createClient method

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

lego-technix committed:
feat(api): configure ready oidc providers on server start

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): use scope constructor parameter

Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): rename authenticationUrlParameters

into extraAuthorizationUrlParameters

Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen committed:
feat(api): use openid client authorizationUrl method

Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7978.review.pix.fr)
  * [App (.org)](https://app-pr7978.review.pix.org)
  * [Orga (.fr)](https://orga-pr7978.review.pix.fr)
  * [Orga (.org)](https://orga-pr7978.review.pix.org)
  * [Certif (.fr)](https://certif-pr7978.review.pix.fr)
  * [Certif (.org)](https://certif-pr7978.review.pix.org)
  * [1D](https://1d-pr7978.review.pix.fr)
  * [Admin](https://admin-pr7978.review.pix.fr)
  * [API](https://api-pr7978.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7978.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7978/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7978/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7978/environment)

reibecca commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
**question:** je ne me souviens plus à quoi correspond le scope ?

lego-technix replied:
Le `scope` est l'ensemble des *claims* auxquels on demande à avoir accès cf. https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

Et le default scope pour Pix API est `openid profile`. Et donc quand Pix API demande un scope = `openid profile` à un OIDC Provider, cela veut dire que Pix API veut savoir d'un utilisateur les propriétés suivantes : 
* `sub`
* `name`, `family_name`, `given_name`, `middle_name`, `nickname`, `preferred_username`, `profile`, `picture`, `website`, `gender`, `birthdate`, `zoneinfo`, `locale`, `updated_at`

reibecca replied:
D'accord merci beaucoup 🙏🏼

reibecca commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
**question:** je ne me souviens plus à quoi correspond le scope ? 
c'est l'app front où est utilisé le sso ?

reibecca commented on api/tests/acceptance/application/authentication/oidc/token-route-post_test.js:
```diff
@@ -76,7 +72,8 @@ describe('Acceptance | Route | oidc | token', function () {
         };
 
-        nock(settings.poleEmploi.tokenUrl).post('/').reply(200, getAccessTokenResponse);
+        //nock('https://oidc.example.net').post('/ea5ac20c-5076-4806-860a-b0aeb01645d4/oauth2/v2.0/token').reply(200, getAccessTokenResponse);
+        oidcExampleNetProvider.client.callback.resolves(getAccessTokenResponse);
```
à supprimer ou à garder ?

lego-technix replied:
Pour expliquer cela j'ai mis une explication (qui était nécessaire, bien vu !) dans la description de la PR : 

> Dans plusieurs fichiers de test on a mis des skip et on a laissé du code en commentaire pour reprendre cette façon de faire plus tard.

lego-technix commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -21,54 +16,59 @@ import { OIDC_ERRORS } from '../../constants.js';
-      postLogoutRedirectUri,
-      additionalRequiredProperties,
-      claimsToStore,
+      wellKnownUrl,
```
À la réflexion à la place de `wellKnownUrl` je préférerais qu'on utilise `openidConfigurationUrl` : 
```suggestion
      openidConfigurationUrl,
```

Je suis désolé, cela a déjà été soulevé une fois par @igorissen, mais je me suis rendu compte en expliquant aux uns et aux autres que c'est vraiment la notion de « openid-configuration » dont il s'agit. Alors je sais ça fait beaucoup de changements, mais je suis prêt à les faire.

lego-technix commented on api/lib/domain/services/authentication/authentication-service-registry.js:
```diff
@@ -5,6 +5,10 @@ class OidcAuthenticationServiceRegistry {
   #readyOidcProviderServicesForPixAdmin = null;
 
+  async configureReadyOidcProviderServices() {
+    await Promise.all(this.#readyOidcProviderServices.map((oidcProviderService) => oidcProviderService.createClient()));
```
:warning: Si un OIDC Provider n'est pas disponible/a des problèmes une erreur est lancée par _openid-client_ qui stoppe le démarrage de Pix API. C'est clairement bloquant.

lego-technix replied:
Je viens de pousser un fixup commit qui modifie `Promise.all` en `Promise.allSettled`.

lego-technix replied:
Je viens de pousser des fixup commits qui modifient `Promise.all` en `Promise.allSettled`, cette dernière fonction ne générant pas de reject si une de ses promises reject et qui indiquent quels OIDC Providers ne sont pas disponibles.

lego-technix reviewed (approved):
✅ Lu et testé fonctionnellement avec succès avec Firefox

igorissen committed:
feat(api): use openid client userinfo method

lego-technix commented:
La définition des nouvelles variables d'environnement en intégration, recette et production a été réalisée pour les 4 SSO actuellement activés en production (« _enabled_ ») : `POLE_EMPLOI`, `CNAV`, `FWB`, `PAYSDELALOIRE`

Je ne me suis pas occupé des variables d’environnement pour le SSO Google.

reibecca commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -323,10 +322,9 @@ class OidcAuthenticationService {
     }
 
     const pickedUserInfo = {
+      sub: userInfo.sub,
```
question: pourquoi on supprime le nonce ?

lego-technix replied:
Il y a 2 raisons dans le même commit. Le commit mériterait d'être séparé en 2 :

lego-technix replied:
Il y a 2 raisons pour lesquelles on supprime le nonce. Et ces 2 suppressions sont effectuées dans le même commit. Le commit mériterait d'être séparé en 2 commits distincts :

* il y avait une erreur dans le code d'avant depuis toujours : il n'y a jamais eu de nonce dans les données renvoyées par le userinfo endpoint
* et enfin on n'a plus besoin de chercher à récupérer le nonce de l'ID Token car la vérification du nonce est faite dans le code suivant : 
```javascript
await this.client.callback(this.redirectUri, { code, state }, { nonce, state: sessionState });
```

lego-technix replied:
@reibecca je viens de séparer le commit supprimant le renvoie du `nonce` par les fonctions OidcAuthenticationService#getUserInfo et OidcAuthenticationService#_getUserInfoFromEndpoint en 2 commits distincts.

Mais le `nonce` est bien utilisé et va être passé jusqu'à la fonction `validateIdToken` de _openid-client_ qui fait le meilleur travail de validation possible :
https://github.com/panva/node-openid-client/blob/main/lib/client.js#L492-L505

lego-technix reviewed (approved):
✅ Relu et retesté fonctionnellement avec succès avec Firefox

lego-technix committed:
fix(api): remove erroneous attempt to get a nonce from the UserInfo endpoint

A UserInfo endpoint never returns a nonce.

This error seems to date back from a long time ago if not from the first
implementation.

lego-technix committed:
refactor(api): remove return of nonce from OidcAuthenticationService#getUserInfo

since it is now useless because the nonce check is now done by the openid-client
library in the following call of Pix:

    await this.client.callback(this.redirectUri, { code, state }, { nonce, state: sessionState });

through the openid-client.Client#validateIdToken function.

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>

EmmanuelleBonnemay reviewed (approved):
Lu et testé fonctionnellement avec succès sur Firefox pour les Fournisseurs d'Identité de la FWB, CNAV, Pays de la Loire. Le test sur France Travail est impossible pour l'instant (problèmes techniques de leur côté).

reibecca reviewed (approved):
![etvRbRt](https://github.com/1024pix/pix/assets/82950611/ab60e928-85e2-43d2-9d0c-adc702c27dd1)
